pkg/cloud/aws/actuators/machine/actuator.go

```diff
@@ -401,49 +202,11 @@ func (a *Actuator) CreateMachine(cluster *clusterv1.Cluster, machine *clusterv1.
 		}
 	}
 
-	userDataEnc := base64.StdEncoding.EncodeToString(userData)
-
-	var iamInstanceProfile *ec2.IamInstanceProfileSpecification
-	if machineProviderConfig.IAMInstanceProfile != nil && machineProviderConfig.IAMInstanceProfile.ID != nil {
-		iamInstanceProfile = &ec2.IamInstanceProfileSpecification{
-			Name: aws.String(*machineProviderConfig.IAMInstanceProfile.ID),
-		}
-	}
-
-	var placement *ec2.Placement
-	if machineProviderConfig.Placement.AvailabilityZone != "" && machineProviderConfig.Subnet.ID == nil {
-		placement = &ec2.Placement{
-			AvailabilityZone: aws.String(machineProviderConfig.Placement.AvailabilityZone),
-		}
-	}
-
-	inputConfig := ec2.RunInstancesInput{
-		ImageId:      amiID,
-		InstanceType: aws.String(machineProviderConfig.InstanceType),
-		// Only a single instance of the AWS instance allowed
-		MinCount:           aws.Int64(1),
-		MaxCount:           aws.Int64(1),
-		KeyName:            machineProviderConfig.KeyName,
-		IamInstanceProfile: iamInstanceProfile,
-		TagSpecifications:  []*ec2.TagSpecification{tagInstance, tagVolume},
-		NetworkInterfaces:  networkInterfaces,
-		UserData:           &userDataEnc,
-		Placement:          placement,
-	}
-
-	runResult, err := client.RunInstances(&inputConfig)
+	instance, err := actuatorec2.LaunchInstance(client, machine.Name, clusterID, machineProviderConfig, userData)
```

Comment:
What's the bar for deciding which goes into the header e.g machineName, cluserId, userdata and what's itemized inside the func e.g ami, sg, subnet...?
I'd be in favour of scoping the header to `(client, machineProviderConfig)` then inside we can itemize by  funcs where relevant e.g `getAMI`, `getUserdata`, etc

Reply:
With just `(client, machineProviderConfig)`, we would need to pass kubernetes client inside as well to collect the user data from a secret. That introduce additional unnecessary dependency on kube client. `machineName` and `clusterID` are here so we don't have to pass entire `Machine` object. Though, I can put it back. So we would end up just with either:
- `LaunchInstance(client awsclient.Client, machineName, clusterID string, machineProviderConfig *providerconfigv1.AWSMachineProviderConfig, userData []byte)`
or
- `LaunchInstance(client awsclient.Client, machineName, clusterID string, machineProviderConfig *providerconfigv1.AWSMachineProviderConfig, kubeClient kubernets.Interface)`
or
- `LaunchInstance(client awsclient.Client, machine *clusterv1.Machine, machineProviderConfig *providerconfigv1.AWSMachineProviderConfig, userData []byte)`
or
- `LaunchInstance(client awsclient.Client, machine *clusterv1.Machine, machineProviderConfig *providerconfigv1.AWSMachineProviderConfig, kubeClient kubernets.Interface)`

Other way is to introduce new data type and put all the clients into it. Though, I don't like using method unless it's necessary.

Reply:
@enxebre I have changed the signature to `LaunchInstance(machine *clusterv1.Machine, lc *LaunchConfig) (*ec2.Instance, error)`. Also wrapping the aws client into a new data type and flipping functions into methods.

Reply:
So the final form of the method is now `LaunchInstance(machine *clusterv1.Machine, machineProviderConfig *providerconfigv1.AWSMachineProviderConfig, userData []byte) (*ec2.Instance, error)`. No LaunchConfig data type.

Reply:
> I'd be in favour of scoping the header to (client, machineProviderConfig) then inside we can itemize by funcs where relevant e.g getAMI, getUserdata, etc cc @frobware

Reply:
> I'd be in favour of scoping the header to (client, machineProviderConfig) then inside we can itemize by funcs where relevant e.g getAMI, getUserdata, etc cc @frobware

Done up to the `getUserdata`, the user data itself are wired through `userData []byte` parameter.